### PR TITLE
Next: Custom queries for the rest CT API methods

### LINKED ISSUE #4865: Custom query for all of api calls inside of commercetools
We have implemented a custom graphql queries mechanism: https://github.com/DivanteLtd/vue-storefront/pull/4808/files but We need to cover by custom queries all of the API connections eg: https://github.com/DivanteLtd/vue-storefront/blob/next/packages/commercetools/composables/src/useCart/index.ts#L17

┆Issue is synchronized with this [Jira Story](https://vsf.atlassian.net/browse/NEXT-277) by [Unito](https://www.unito.io/learn-more)

### PR DESCRIPTION
**Pull request by lukasborawski**
### Related Issues

Closes #4865 

### Short Description and Why It's Useful

Add the ability to provide custom queries for the rest of CT API methods.

### Which Environment This Relates To
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog
- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing Vue Storefront sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and Currently Important Rules Acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# editorconfig.org
```
Comment (andrzejewsky) [inline .editorconfig:1]: I think you committed this accidentally

File: package.json [hunk 2]
```diff
@@ -86,6 +86,7 @@
     "jest": "^24.9.0",
     "lerna": "^3.15.0",
     "lint-staged": "^10.0.7",
+    "prettier": "^1.16.4",
```
Comment (andrzejewsky) [inline package.json:89]: that's a good idea, but not i the scope of this issue, probably that's the reason of breaking build (linter) - if this is not a problem, could you introduce pretier as a separated PR? but if you are able to fix all of build issues quickly - it's fine

File: packages/commercetools/api-client/src/api/addToCart/index.ts [hunk 3]
```diff
@@ -1,14 +1,23 @@
...
-    version: cart.version,
-    actions: [createAddLineItemAction(product, quantity)]
-  });
+const addToCart = async (
+  { id, version }: Cart,
+  product: ProductVariant,
+  quantity: number,
+  customQuery?: CustomQueryFn
+): Promise<CartResponse> => {
+  console.log(id);
```
Comment (andrzejewsky) [inline packages/commercetools/api-client/src/api/addToCart/index.ts:12]: whooopssss

File: packages/commercetools/api-client/src/types/Api.ts [hunk 4]
```diff
@@ -7,9 +7,21 @@ export interface CustomQuery<T> {
...
+  cart?: CustomQuery<T>,
+  user?: CustomQuery<T>
+) => { cart: CustomQuery<T>; user: CustomQuery<T> };
 
-export const getCustomQuery = <T = any>(customQueryFn: CustomQueryFn<T>, defaultQuery) => customQueryFn ? customQueryFn() : { query: defaultQuery, variables: {} };
+export const getCustomQuery = <T = any>(customQueryFn: CustomQueryFn<T>, defaultQuery) =>
+  customQueryFn ? customQueryFn() : { query: defaultQuery, variables: {} };
+export const getCartCustomQuery = <T = any>(customQueryFn: CartCustomQueryFn<T>, customQueries) =>
+  customQueryFn
+    ? customQueryFn()
```
Comment (andrzejewsky) [inline packages/commercetools/api-client/src/types/Api.ts:20]: You have to pass to this function the default queries as sometimes you would need to override variables or only one query.

```
cutomQuery({ cart, user }) => {
  // cart and user  has default queries here
  return {
    cart,
    user: { query: user.query, variables: { I'm changing just variables}  }
  }
} 
```

File: packages/commercetools/composables/src/useCart/currentCart.ts [hunk 5]
```diff
@@ -1,15 +1,10 @@
...
-    return profileResponse.data.me.activeCart;
-  }
-
-  const cartResponse = await createCart();
-
-  return cartResponse.data.cart;
+const loadCurrentCart = async (customQuery) => {
+  const { data: profileData } = await getMe(customQuery);
+  if (profileData.me.activeCart) return profileData.me.activeCart;
+  const { data } = await createCart(customQuery);
```
Comment (andrzejewsky) [inline packages/commercetools/composables/src/useCart/currentCart.ts:6]: don't pass same function to both api function, just resolve queries and pass them

```
const { user, cart } = customQuery(...)

const { data: profileData } = await getMe(user);
....
 const { data } = await createCart(cart);
```

Comment (andrzejewsky) [reply packages/commercetools/composables/src/useCart/currentCart.ts:6]: not both API function doesn't know about each other's queries

**Review by andrzejewsky — CHANGES_REQUESTED** [review]

File: packages/commercetools/api-client/src/types/Api.ts [hunk 6]
```diff
@@ -7,9 +7,21 @@ export interface CustomQuery<T> {
...
+    : {
+      cart: customQueries.cart ? { query: customQueries.cart.query, variables: {} } : null,
```
Comment (filrak) [inline packages/commercetools/api-client/src/types/Api.ts:22]: I remember we talked about naming for this example to be more descriptive, are there any other examples that need 2 queries/mutations?

File: .editorconfig (hunk 1 above)
Comment (lukasborawski) [reply .editorconfig:1]: For this configuration yes. But why we all can't have the same formatting? The content might be under discussion.

File: packages/commercetools/api-client/src/types/Api.ts (hunk 6 above)
Comment (lukasborawski) [reply packages/commercetools/api-client/src/types/Api.ts:22]: @filrak Not yet.

File: package.json (hunk 2 above)
Comment (lukasborawski) [reply package.json:89]: @andrzejewsky I've fixed the eslint tracked errors.

File: .editorconfig (hunk 1 above)
Comment (lukasborawski) [reply .editorconfig]: Removed.

File: package.json (hunk 2 above)
Comment (lukasborawski) [reply package.json]: Well... Actually I removed it. Will create separate PR for that. 🙈

File: packages/commercetools/api-client/src/types/Api.ts [hunk 7]
```diff
@@ -7,9 +7,20 @@ export interface CustomQuery<T> {
   variables: T;
 }
 
-type CustomQueryFn<T = any> = (query?: any, variables?: T) => CustomQuery<T>
+export type CustomQueryFn<T = any> = (query?: any, variables?: T) => CustomQuery<T>;
 
-export const getCustomQuery = <T = any>(customQueryFn: CustomQueryFn<T>, defaultQuery) => customQueryFn ? customQueryFn() : { query: defaultQuery, variables: {} };
+/* export type CartCustomQueryFn<T> = (
```
Comment (andrzejewsky) [inline packages/commercetools/api-client/src/types/Api.ts:12]: do you need it?

File: packages/core/core/src/types.ts [hunk 8]
```diff
@@ -111,23 +111,44 @@ export interface UseCategory
...
+export function RemoveFromCart<CART_ITEM>(product: CART_ITEM, customQuery?: {}): any {
+  return { product, customQuery };
+}
+export function ApplyCoupon(coupon): Promise<void>
+export function ApplyCoupon(coupon, customQuery: {}): Promise<void>
+export function ApplyCoupon(coupon, customQuery?: {}): any {
+  return { coupon, customQuery };
+}
+export function LoadCart(customQuery?: {}): Promise<void>
+export function LoadCart(customQuery?: {}): any {
```
Comment (andrzejewsky) [inline packages/core/core/src/types.ts:128]: @filrak look at this, he solved the problem with overloading functions using TS, actually I really like this - this case, if some platform doesn't have a custom query, you can easily handle this - we had similar problems in the past, remember? and we actually postponed solving that.

@lukasborawski can you create and issue to refactor other core interfaces in this way?
Let's say that goal could be - move each TS function definition under the composable interfaces to the separated function like you did. I'm pretty sure we'll find other places where overloading would be usefull

**Review by andrzejewsky — APPROVED** [review]

File: packages/commercetools/api-client/src/types/Api.ts (hunk 7 above)
Comment (lukasborawski) [reply packages/commercetools/api-client/src/types/Api.ts:12]: No. Removing.

File: packages/core/core/src/types.ts (hunk 8 above)
Comment (lukasborawski) [reply packages/core/core/src/types.ts:128]: @andrzejewsky I believe that @filrak already knows that. We spoke about this feature before.

**Review by filrak — APPROVED** [review]

**Pull request merged**